File: python/stack.py
# A stack is a linear data structure that follows the Last In, First Out (LIFO)
# principle, meaning that the most recently added element is the first to be
# removed. The main operations associated with a stack are push
# (to add an element), pop (to remove the top element), peek
# (to view the top element without removing it)

import random


class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.next = None


class Stack:
    def __init__(self) -> None:
        self.__head = None
        self.length = 0

    def push(self, data) -> None:
        new_node = Node(data)

        self.length += 1

        if not self.__head:
            self.__head = new_node
            return

        new_node.next = self.__head
        self.__head = new_node

    def pop(self):
        if not self.__head:
            return

        # Either keep it at 0, or decrement if not negative
        self.length = max(0, self.length - 1)

        current_head = self.__head

        if self.length == 0:
            self.__head = None
        else:
            self.__head = current_head.next

        return current_head.data

    def peek(self):
        return self.__head.data if self.__head else None

    def __str__(self) -> str:
        msg = f"Length: {self.length} "

        current_head = self.__head

        while current_head:
            msg += f"[{current_head.data}] -> "
            current_head = current_head.next

        msg += "END"
        return msg


stack = Stack()

print(stack)

stack.push(3)

print(stack)

stack.push(7)

print(stack)

print(f"Removed: {stack.pop()}")

print(stack)

for i in range(0, 5):
    stack.push(int(random.random() * 100))

print(stack)

print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")
print(f"Removed: {stack.pop()}")

File: typescript/stack.ts
import { strict as assert } from "node:assert";

type TNode<T> = {
    data: T;
    next?: TNode<T>;
};

class Stack<T> {
    public length: number;
    private head: TNode<T> | undefined;

    constructor() {
        this.length = 0;
        this.head = undefined;
    }

    push(data: T): void {
        const node = { data: data } as TNode<T>;

        if (this.length === 0) {
            this.head = node;
            this.length++;
            return;
        }

        this.length++;

        node.next = this.head;
        this.head = node;
    }

    pop(): T | undefined {
        if (this.length === 0) return;

        this.length = Math.max(0, --this.length);

        const head = this.head;

        if (this.length === 0) {
            this.head = undefined;
            delete head?.next;
            return head?.data;
        }

        this.head = this.head?.next;
        delete head?.next;
        return head?.data;
    }

    peek(): T | undefined {
        return this.head?.data;
    }
}

const stack = new Stack<number>();

stack.push(5);
stack.push(7);
stack.push(9);

assert(stack.pop() === 9, "Expected 9 to be popped");
assert(stack.length === 2, "Expected stack length to be 2");

stack.push(11);
assert(stack.pop() === 11, "Expected 11 to be popped");
assert(stack.pop() === 7, "Expected 7 to be popped");
assert(stack.peek() === 5, "Expected 5 to be on top of the stack");
assert(stack.pop() === 5, "Expected 5 to be popped");
assert(
    stack.pop() === undefined,
    "Expected popping from an empty stack to return undefined",
);

stack.push(69);
assert(stack.peek() === 69, "Expected 69 to be on top of the stack");
assert(stack.length === 1, "Expected stack length to be 1");

console.log("ALL TESTS PASSED!");
